integration-tests/__tests__/expect_async_matcher.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+'use strict';
+
+async function toHaveLengthAsync(
+  received: any,
+  lengthPromise: Promise<number>,
+) {
+  const length = await lengthPromise;
+
+  const pass = received.length === length;
+  const message = pass
+    ? () =>
+        `Expected value to not have length:\n` +
+        `  ${length}\n` +
+        `Received:\n` +
+        `  ${received}\n` +
+        `received.length:\n` +
+        `  ${received.length}`
+    : () =>
+        `Expected value to have length:\n` +
+        `  ${length}\n` +
+        `Received:\n` +
+        `  ${received}\n` +
+        `received.length:\n` +
+        `  ${received.length}`;
+
+  return {message, pass};
+}
+
+expect.extend({
+  toHaveLengthAsync,
+});
+
+it('works with expected non promise values', async () => {
+  await (expect([1]): any).toHaveLengthAsync(Promise.resolve(1));
+});
+
+it('works with expected non promise values is a failure', async () => {
+  try {
+    await (expect([1]): any).toHaveLengthAsync(Promise.resolve(2));
+  } catch (error) {
+    expect(error).toBeDefined();
+    expect(error.message).toMatchSnapshot();
+  }
+});
+
+it('works with expected non promise values and not', async () => {
+  await (expect([1, 2]): any).not.toHaveLengthAsync(Promise.resolve(1));
+});
+
+it('works with expected non promise values and not is a failure', async () => {
+  try {
+    await (expect([1, 2]): any).not.toHaveLengthAsync(Promise.resolve(2));
+  } catch (error) {
+    expect(error).toBeDefined();
+    expect(error.message).toMatchSnapshot();
+  }
+});
+
+it('works with expected promise values', async () => {
+  await (expect(Promise.resolve([1])).resolves: any).toHaveLengthAsync(
+    Promise.resolve(1),
+  );
+});
+
+it('works with expected promise values is a failure', async () => {
+  try {
```

Comment:
I don't think it's necessary to have the `catch` here - just let the test fail and assert on the output from the integration test

Reply:
@SimenB Sorry, what output are you referring to ? Maybe there is a way to get the output that I'm not aware of. Thanks for the time BTW!

Reply:
In the integration test, you have access to `stderr` and `stderr`. You can assert on the output of jest from there

Reply:
See e.g. https://github.com/facebook/jest/blob/master/integration-tests/__tests__/failures.test.js

Reply:
Got it! Will change the tests to use `stderr`. Thanks!